### PR TITLE
Sync: improve importer detection and sync start action

### PR DESCRIPTION
**Pull request by jonathansadowski**
Take two on the work started in #11523. See D25761-code.

#### Changes proposed in this Pull Request:

* Adds a utility function (`get_calling_importer_class`) which infers the import class responsible for calling the import action.
* Reworks the logic for triggering the sync action so that it is the same for `import_start`, `import_done`, and `import_end`.

#### Testing instructions:

* Go to `/wp-admin/admin.php?import=wordpress`
* Complete an import. If the import completes successfully, two actions should be synced
  * `jetpack_sync_import_start`
  * `jetpack_sync_import_end`
* Both events should have two parameters (`wordpress`, `WordPress`)
* Other types of imports that trigger the `import_done` or `import_end` actions should also sync `jetpack_sync_import_done`. Examples: (Blogger, LiveJournal, RSS). Each should have two parameters, and as long as the importer is in `known_importers`, those parameters should reflect those values.  These other importers, however, don't currently trigger `import_start`, so `jetpack_sync_import_start` won't be triggered.

#### Proposed changelog entry for your changes:

No changelog entry needed

### CODE REVIEW

File: sync/class.jetpack-sync-module-import.php
Comment (jblz) [inline sync/class.jetpack-sync-module-import.php]: It may be worth it to comment as to why `import_done` & `import_end` both map to `jetpack_sync_import_end`

File: sync/class.jetpack-sync-module-import.php
Comment (jblz) [inline sync/class.jetpack-sync-module-import.php]: If some request triggers both `import_done` & `import_end` it looks like this would limit it to a single `jetpack_sync_import_end`, correct?  That seems like the desired behavior here, so 👍 if so :)

Is there any instance where these would be triggered in separate request cycles for the same import?  If so, is it acceptable that we'd have repeat sync events?  Probably not a huge issue if so.

**Review by jblz — COMMENTED** [review]
This looks good and is working well for me, but will hold off marking my approval until the "synced wpcom file" issue is resolved. I left a couple of questions in the meantime.

File: sync/class.jetpack-sync-module-import.php
Comment (jonathansadowski) [reply sync/class.jetpack-sync-module-import.php]: Thanks for the suggestion. I just pushed a commit that adds a comment explaining why they map to the same sync action.

File: sync/class.jetpack-sync-module-import.php
Comment (jonathansadowski) [reply sync/class.jetpack-sync-module-import.php]: That is correct. The replaced code worked similarly (though, it only handled a single action mapping, ensuring `jetpac_sync_import_end` was only triggered once). The update 

While I suppose that it's possible there is an importer that would trigger multiple `import_done` or `import_end` actions outside of a single request, none of the primary importers that we're syncing currently do. If we find this to be the case, we can certainly revisit -- but since it's something that importers _shouldn't_ do, until and unless we see it in practice, I think it's acceptable.

**Comment by jonathansadowski** [comment]
Thanks for the review @jblz. It looks like the process of pushing my previous commit caused the block by p9dueE-wa-p2 to no longer apply. I'll review D25918-code.

File: sync/class.jetpack-sync-module-import.php
Comment (jblz) [inline sync/class.jetpack-sync-module-import.php]: Is it worth guarding against implementation changes in `wp_debug_backtrace_summary`, etc. by allowing for optional whitespace around the argument in `do_action`?

Something like this:

```suggestion
			if ( preg_match( "/^do_action\s*\\(\s*\'([^\']+\s*)/", $backtrace[ $i ], $matches ) ) {
```

Comment (jonathansadowski) [reply sync/class.jetpack-sync-module-import.php]: Thanks, great idea -- I just committed your suggestion!

**Review by jblz — DISMISSED** [review]
This is working well! Accepting on behalf of Samus Team.

**Comment by kraftbj** [comment]
Asking for a set of Poseidon eyes. The changes look reasonable to me, though there are some new PHPCS issues added. I'd prefer that the changed lines don't introduce new issues, but that isn't a blocker for this file yet.

**Comment by jonathansadowski** [comment]
Thanks for adding @Automattic/poseidon to the review @kraftbj!

I've updated the PR to address the PHPCS issues.

File: sync/class.jetpack-sync-module-import.php
Comment (lezama) [inline sync/class.jetpack-sync-module-import.php]: have you seen https://github.com/Automattic/jetpack/blob/dd9c8a374f4a136e4499853becb124b4270561d3/class.jetpack.php#L6965 maybe we could extend that one or add a new method to Jetpack.class

File: sync/class.jetpack-sync-module-import.php
Comment (enejb) [inline sync/class.jetpack-sync-module-import.php]: It would be really great to leave a comment here. So when someone searched for `jetpack_sync_import_start` or `jetpack_sync_import_end` we also get this line.

File: sync/class.jetpack-sync-module-import.php
Comment (lezama) [inline sync/class.jetpack-sync-module-import.php]: we could use [Sync Settings](https://github.com/Automattic/jetpack/blob/master/sync/class.jetpack-sync-settings.php) here, the good thing about it is that they are easy to update remotely or via cli if one day we want to add a new importer to the list without waiting for a release.

**Comment by kraftbj** [comment]
Marking as needing @jonathansadowski's responses. Note that Monday EOD is code freeze for 7.3, so love to get this wrapped up and in the next couple of business days.

File: sync/class.jetpack-sync-module-import.php
Comment (creativecoder) [reply sync/class.jetpack-sync-module-import.php]: I've updated the PR to use a default sync setting for `known_importers`, using the conventions I found in `Jetpack_Sync_Settings` and `Jetpack_Sync_Defaults`

File: sync/class.jetpack-sync-module-import.php
Comment (jonathansadowski) [reply sync/class.jetpack-sync-module-import.php]: Thanks for the suggestion @lezama. I wasn't aware of that method prior to creating `get_calling_importer_class`. I explored creating something similar in `Jetpack.class`, but there are a few issues with such an approach:

* The existing method simply returns _if_ any of an array of names is a function in the backtrace. It'd be pretty trivial to extend that to support any column of the backtrace (not just `function`, but also `class`), but we care about _which_ class is found in the backtrace.
* We currently care not only about known importers, but any class that extends `WP_Importer`. We could of course add logic similar to the logic we have in `get_calling_importer_class` that checks for subclasses, but that would undo many of the optimizations currently in place in `is_function_in_backtrace`
* All of these extra needs are really very specific to our implementation of getting the calling class in Sync for importers. Until we have the need to do something similar outside of Jetpack Sync, I'm hesitant to write something more generalized without a better idea of how it'd be used.

File: sync/class.jetpack-sync-module-import.php
Comment (jonathansadowski) [reply sync/class.jetpack-sync-module-import.php]: Great suggestion. Comment added!

**Comment by creativecoder** [comment]
@kraftbj @lezama I believe we've addressed all the outstanding comments if it's still possible to get this in for 7.3 🤞.

File: sync/class.jetpack-sync-module-import.php
Comment (kraftbj) [inline sync/class.jetpack-sync-module-import.php]: I think this is going to trip up our developer.jetpack.com parser, or at least not be able to parse it.

@jeherve Do you have ideas of how to document this? We could add the two `do_action` calls in another file that is never called for the sake of documentation, linking back to here as the actual place it fires? I'm okay adding that after merging to not block this so we can get to freeze.

Comment (jonathansadowski) [reply sync/class.jetpack-sync-module-import.php]: Since there are only 2 possible `$sync_action`s, we could also just use a conditional or a switch statement here...

```
switch( $sync_action ) {
	case 'jetpack_sync_import_start':
		do_action( 'jetpack_sync_import_start' ... );
		break;
	...
}
```

Comment (jeherve) [reply sync/class.jetpack-sync-module-import.php]: Yeah, let's go with a conditional. It's not super pretty, but it should make the parser happy and may actually end up being easier to get back to later, when looking for specific actions for example.

Comment (kraftbj) [reply sync/class.jetpack-sync-module-import.php]: Thank you both. Let's do the switch and add inline documentation above each do_action. @jonathansadowski do you have bandwidth to knock that out now-ish? If not, I can get to it in a short bit.

Comment (jonathansadowski) [reply sync/class.jetpack-sync-module-import.php]: Yep, grabbing it right now. Thanks @jeherve & @kraftbj.

**Review by kraftbj — APPROVED** [review]

**Pull request merged**
